src/main/kotlin/com/sainsburys/k2zpl/command/ZplCommand.kt

```diff
@@ -1,9 +1,15 @@
 package com.sainsburys.k2zpl.command
 
-interface ZplCommand {
-    val command: CharSequence
-    val parameters: Map<CharSequence, Any?> get() = addParameters()
-    fun build(stringBuilder: StringBuilder) = stringBuilder.apply {
+import java.util.Collections.unmodifiableMap
+
+abstract class ZplCommand(
+    parameters: List<Pair<CharSequence, Any?>> = emptyList()
+) {
+    val parameters: Map<CharSequence, Any?> = unmodifiableMap(parameters.toMap())
```

Comment:
As alluded to in the OP, this use of `unmodifiableMap` is what I'm most ambivalent about, as on one hand it stops `parameters` being casted into a mutable map, but on the other hand it's not idiomatic in Kotlin.

What do you think?

Reply:
Hi @jbduncan,

I took at look at this and I think it made me realise we don't have to use a raw `Map` as the parameters here - we can just wrap it in another class where we have full control of the implementation and usage. Have a look at #21

Reply:
Nice! I'm closing this PR in favour of yours then. :)